File: SWEA/Unrated/5658. ［모의 SW 역량테스트］ 보물상자 비밀번호/［모의 SW 역량테스트］ 보물상자 비밀번호.java
import java.util.*;

public class Solution {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        for (int tc = 1; tc <= t; tc++) {
            int N = scanner.nextInt();
            int K = scanner.nextInt();
            int answer = 0;
            String arr = scanner.next();
            List<Character> lst = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                lst.add(arr.charAt(i));
            }
            List<Integer> answer_lst = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                for (int m = 0; m < N; m += N / 4) {
                    String temp = "";
                    for (int j = m; j < m + N / 4; j++) {
                        temp += lst.get(j);
                    }
                    int int_temp = Integer.parseInt(temp, 16);
                    answer_lst.add(int_temp);
                }
                lst.add(0, lst.get(lst.size() - 1));
                lst.remove(lst.size() - 1);
            }
            Set<Integer> answer_set = new HashSet<>(answer_lst);
            List<Integer> a_s = new ArrayList<>(answer_set);
            Collections.sort(a_s, Collections.reverseOrder());
            answer = a_s.get(K - 1);
            System.out.println("#" + tc + " " + answer);
        }
    }
}

File: SWEA/Unrated/5658. ［모의 SW 역량테스트］ 보물상자 비밀번호/［모의 SW 역량테스트］ 보물상자 비밀번호.py
t = int(input())
for tc in range(1, t+1):
    N, K = map(int, input().rstrip().split())
    answer = 0
    arr = input().rstrip()
    lst = []
    for i in range(N):
        lst.append(arr[i])
    answer_lst = []
    for i in range(N):
        for m in range(0, N, N//4):
            temp = "".join(lst[m:(m+N//4)])
            int_temp = int(temp, 16)
            answer_lst.append(int_temp)
        lst.insert(0, lst[-1])
        del lst[-1]

    answer_set = set(answer_lst)
    a_s = sorted(answer_set, reverse=True)
    answer = a_s[K-1]
    print(f"#{tc} {answer}")